### PR TITLE
WIP: Update test_to_records to test with `lengths` argument, added co…

### PR DESCRIPTION
**Pull request by asmith26**
…rresponding functionality based on `.to_dask_array()`

- [ ] Tests added / passed
- [ ] Passes `flake8 dask`

### CODE REVIEW

**Comment by mrocklin** [comment]
cc @TomAugspurger

File: dask/dataframe/core.py [hunk 1]
```diff
@@ -3227,9 +3227,34 @@ def pivot_table(self, index=None, columns=None,
         return pivot_table(self, index=index, columns=columns, values=values,
                            aggfunc=aggfunc)
 
-    def to_records(self, index=False):
+    def to_records(self, lengths=None, index=False):
```
Comment (TomAugspurger) [inline dask/dataframe/core.py]: I think lengths should be the last argument. That way we match the pandas signature more closely (though we're missing other kwargs currently).

File: dask/dataframe/core.py [hunk 2]
```diff
@@ -3227,9 +3227,34 @@ def pivot_table(self, index=None, columns=None,
...
         from .io import to_records
-        return to_records(self)
+        from dask.array.core import normalize_chunks
+
+        if lengths is True:
```
Comment (TomAugspurger) [inline dask/dataframe/core.py]: I've been regretting the use of `lengths=True`. I fear that it's not explicit enough.

What do people think about deprecating `lengths=True` and requiring something like `lengths='compute'` to make it clear what you're doing?

File: dask/dataframe/core.py [hunk 3]
```diff
@@ -3227,9 +3227,34 @@ def pivot_table(self, index=None, columns=None,
...
+            lengths = tuple(self.map_partitions(len).compute())
+
+        records = to_records(self)
+
+        if isinstance(lengths, Sequence):
```
Comment (TomAugspurger) [inline dask/dataframe/core.py]: Move this to a `_validate_chunks` method and call like

```
chunks = self._validate_chunks(records, lengths)
records._chunks = chunks
```

When `lengths` is None, you'll return `records._chunks` from `_validat_chunks`.

and call from `to_dask_array` as well.

File: dask/dataframe/io/tests/test_io.py [hunk 4]
```diff
@@ -496,15 +496,29 @@ def test_to_bag():
...
     from dask.array.utils import assert_eq
     df = pd.DataFrame({'x': ['a', 'b', 'c', 'd'],
                        'y': [2, 3, 4, 5]},
                       index=pd.Index([1., 2., 3., 4.], name='ind'))
     ddf = dd.from_pandas(df, 2)
 
-    assert_eq(df.to_records(), ddf.to_records())
+    assert_eq(df.to_records(), ddf.to_records(lengths=lengths))
+
+    # assert isinstance(result, da.Array)
```
Comment (TomAugspurger) [inline dask/dataframe/io/tests/test_io.py]: Was this failing if not commented out?

File: dask/dataframe/io/tests/test_io.py [hunk 5]
```diff
@@ -496,15 +496,29 @@ def test_to_bag():
...
+    #
+    # expected_chunks = ((2, 3),)
+    #
+    # if as_frame:
+    #     expected_chunks = expected_chunks + ((1,),)
+    #
+    # assert result.chunks == expected_chunks
```
Comment (TomAugspurger) [inline dask/dataframe/io/tests/test_io.py]: Add some tests for the cases that raise as well

1. len(lenghts) != npartitions
2. an invalid value for lengths.

File: dask/dataframe/core.py (hunk 2 above)
Comment (asmith26) [reply dask/dataframe/core.py]: I agree `lengths='compute'` is clearer. Will not add to this commit for now - may be worth creating a new issue to change all occurrences (e.g. with `.to_dask_array()`) + docs etc. ?

File: dask/dataframe/core.py (hunk 3 above)
Comment (asmith26) [reply dask/dataframe/core.py]: Would we need to test the new `_validate_chunks()` method?

File: dask/dataframe/io/tests/test_io.py (hunk 4 above)
Comment (asmith26) [reply dask/dataframe/io/tests/test_io.py]: Sorry you are correct, this does not fail when not commented.

File: dask/dataframe/io/tests/test_io.py (hunk 5 above)
Comment (asmith26) [reply dask/dataframe/io/tests/test_io.py]: In progress (hopefully this week, possibly next).

File: dask/dataframe/core.py [hunk 6]
```diff
@@ -3248,9 +3248,34 @@ def pivot_table(self, index=None, columns=None,
         return pivot_table(self, index=index, columns=columns, values=values,
                            aggfunc=aggfunc)
 
-    def to_records(self, index=False):
+    def to_records(self, index=False, lengths=None):
```
Comment (asmith26) [inline dask/dataframe/core.py]: Reordered arguments following https://github.com/dask/dask/pull/4515#discussion_r265154779

Although index is not currently being used? Should it be added to https://github.com/asmith26/dask/blob/c0cf9f7c1717d7ec40fbda4d6cebbc4421e3d137/dask/dataframe/core.py#L3264 ?

**Comment by asmith26** [comment]
Hi @TomAugspurger 

Thanks for your feedback! I think I've completed all of your requests.

The tests generally mirror the *to_dask_array* tests and  pass, however I'm a bit concerned that I have commented out the check for equality: https://github.com/asmith26/dask/blob/issue-4469/dask/dataframe/io/tests/test_io.py#L521

If I uncomment this I get a Key error still- I understand what the error is, but I am not familiar enough with Dask internals to understand why the key computed by Dask is different to that in the graph; in particular:

This key
```python
key = ('to_records-7efaccf97358a6d0b69926d9bf624a17', 0, 0)
```

Is trying to index the dict:
```python
dsk = {('from_pandas-b581c7965fed6d41f34e734ae95a309c', 0):      x  y
ind      
1.0  a  2
2.0  b  3, ('from_pandas-b581c7965...ubgraph_callable, ('from_pandas-b581c7965fed6d41f34e734ae95a309c', 1), 'from_pandas-b581c7965fed6d41f34e734ae95a309c')}
```
which clearly doesn't exist.

It appears to be getting the key within: https://github.com/asmith26/dask/blob/master/dask/base.py#L206-L207

Any thoughts?

**Comment by asmith26** [comment]
I'll commit with check for equality below to make it clearer:

File: dask/dataframe/core.py [hunk 7]
```diff
@@ -1849,7 +1832,31 @@ def values(self):
...
+            else:
+                chunks = normalize_chunks((lengths, (len(self.columns),)))
+
+            return chunks
+        elif lengths is not None:
+            raise ValueError("Unexpected value for 'lengths': '{}'".format(lengths))
+
+        return arr._chunks
+
+      def _is_index_level_reference(self, key):
```
Comment (TomAugspurger) [inline dask/dataframe/core.py]: Formatting issue I think (too many spaces?)

File: dask/dataframe/io/tests/test_io.py [hunk 8]
```diff
@@ -505,6 +505,42 @@ def test_to_records():
...
+                       'y': [2, 3, 4, 5]},
+                      index=pd.Index([1., 2., 3., 4.], name='ind'))
+    ddf = dd.from_pandas(df, 2)
+
+    result = ddf.to_records(lengths=lengths)
+    assert_eq(df.to_records(), result)
+
+    assert isinstance(result, da.Array)
+
+    expected_chunks = ((2, 2), (2,))
```
Comment (TomAugspurger) [inline dask/dataframe/io/tests/test_io.py]: This needs to be `((2, 2),)`.

File: dask/dataframe/core.py [hunk 9]
```diff
@@ -3295,9 +3302,18 @@ def pivot_table(self, index=None, columns=None,
...
         from .io import to_records
-        return to_records(self)
+
+        if lengths is True:
+            lengths = tuple(self.map_partitions(len).compute())
+
+        records = to_records(self)
+
+        chunks = self._validate_chunks(records, lengths)
+        records._chunks = chunks
```
Comment (TomAugspurger) [inline dask/dataframe/core.py]: I'm not very familiar with record arrays, but apparently they're 1-dimensional?

```python
In [6]: df = pd.DataFrame({"A": [1, 2], "B": [3, 4]})

In [7]: df.to_records().ndim
Out[7]: 1
```

So this should (maybe) be `records._chunks = (chunks[0],)`.

**Comment by jrbourbeau** [comment]
I pushed a commit that addresses your review comments @TomAugspurger. Tests are passing now, so a quick re-review whenever you get a chance would be appreciated.

**Comment by TomAugspurger** [comment]
Thanks @asmith26 and @jrbourbeau!

**Comment by asmith26** [comment]
Thank you very much for everyone's help! I thoroughly enjoyed tackling this, and I've learned loads from everyone's feedback.

Hope I'm able to help out with this brilliant library in the future! :)

**Pull request merged**